### ISSUE TITLE
Rename Feature Corrupts Unsaved Files in VS2017 15.7.1 

### ISSUE THREAD

**Issue by Zagrophyte** [Microsoft/visualfsharp#4927]
When you rename a symbol that exists in another file that is open with pending changes, the rename operation is computed using the character positions of the _unsaved_ file. This means if you've added _any_ lines or changed target symbol positions in the target files, the replace operation will overwrite any text that existed at the old offsets.

UPDATE: While writing this issue I have now also encountered the crashing issues when renaming symbols added and used within the same unsaved file. (seems related to #4655 and #2204).

#### Repro steps

1. Open a source file which contains a symbol that is defined in another source file.
![image](https://user-images.githubusercontent.com/6778687/40149957-d8b4a220-5944-11e8-91b5-ddb6bf939200.png)

2. Add a few lines of code _above_ where that symbol is used. **DO NOT SAVE YET**
![image](https://user-images.githubusercontent.com/6778687/40149992-0de3eba4-5945-11e8-949d-62e936c26571.png)

3. Go to the definition of that symbol in the other file (F12). and use the Rename feature (Ctrl+R+R) to rename that symbol.
![image](https://user-images.githubusercontent.com/6778687/40150041-55599538-5945-11e8-8f2d-a6f2f23018cb.png)

4. The Rename operation overwrites code above where the actual symbols were used and saves the file, corrupting the source code.
![image](https://user-images.githubusercontent.com/6778687/40150051-6225d290-5945-11e8-8199-355fe70f3ed7.png)

#### Bonus Content
Interestingly, I tried it again with Preview Changes enabled, and the resulting preview window actually shows the corruption:
![image](https://user-images.githubusercontent.com/6778687/40150656-de586924-5948-11e8-9558-387a3e6e85df.png)

#### Repro Solution ZIP
[VS2017.15.7.1-BlindRenameOffsetIssue.zip](https://github.com/Microsoft/visualfsharp/files/2011226/VS2017.15.7.1-BlindRenameOffsetIssue.zip)

#### Expected behavior

Rename feature properly renames usages in open / modified files.

#### Actual behavior

Rename feature uses old buffer to calculate rename operations, overwriting current buffer and saving the file.

#### Known workarounds

Only to save all open files before performing any rename operations.

#### Related information 

Provide any related information 

* Operating system
    * Windows 10, Version 1709 (16299.309)
* Branch
    * 15.7.1 public release
* .NET Runtime, CoreCLR or Mono Version
    * .NET Framework 4.7.1
* Editing Tools (e.g. Visual Studio Version)
    * Visual Studio 15.7.1
* Indications of severity
    * This issue and the related Rename->VSCrash issues can cause data loss.

**Comment by vasily-kirichenko** [comment]
Renaming does not work on dirty buffers, by design.

**Comment by Zagrophyte** [comment]
If that’s true there are two ways to address this: 1) Force a save before starting the rename operation. 2) Prompt the user to do so before allowing it to continue.

I lean towards #1.

**Comment by cartermp** [comment]
Closing this in favor of #2204 as the canonical issue
